### ISSUE TITLE
null pointer after release 1.3.6

### ISSUE THREAD

**Issue by jburse** [jburse/jekejeke-devel#3]
From stan,

I am getting a null reference exception in the code below. To reproduce, consult the code and do

```
?- run42a.

% Busy Beaver Turing machine.
% Author: Stanislav Maslovski. License CC BY 4.0

machine22("1RB1LB1LA1RH").

machine32("1RB1RH1LB0RC1LC1LA").

machine42a("1RB1LB1LA0LC1RH1LD1RD0RA").

machine42b("1RB0RC1LA1RA1RH1RD1LD0LB").

machine42c("1RB0RD1LC0LA1RA1LB1RH0RC").


:- dynamic(trans/4).

run22 :-
    machine22(M),
    prepare(M),
    turing(0'A,([0,0],0,[0]),_,_).

run32 :-
    machine32(M),
    prepare(M),
    turing(0'A,([0],0,[0,0,0]),_,_).

run42a :-
    machine42a(M),
    prepare(M),
    turing(0'A,([0,0,0,0,0,0,0,0,0,0],0,[0,0,0]),_,_).

run42b :-
    machine42b(M),
    prepare(M),
    turing(0'A,([0,0],0,[0,0,0,0,0,0,0,0,0,0,0,0]),_,_).

run42c :-
    machine42c(M),
    prepare(M),
    turing(0'A,([0,0,0,0,0],0,[0,0,0,0,0,0,0]),_,_).

prepare(M) :-
    abolish(trans/4),
    parse(0, 0'A, M).   

parse(_, _, []) :- !.
parse(H, S, [X,RL,S1|Ds]) :-
    H1 is X-0'0,
    ( RL = 0'R -> B = wsr(H,H1,T,T1)
    ; RL = 0'L -> B = wsl(H,H1,T,T1) ),
    assertz((trans(S,T,S1,T1) :- B, !)),
    ( H = 0 -> parse(1, S, Ds)
    ; H = 1 -> N is S+1, parse(0, N, Ds) ). 

turing(0'H, T, 0'H, T) :- !.
turing(S, T, S2, T2) :-
    trans(S, T, S1, T1),
    dump(S1, T1),
    turing(S1, T1, S2, T2).

wsr(H, H1, (L,H,[X|R]), ([H1|L],X,R)).

wsl(H, H1, ([X|L],H,R), (L,X,[H1|R])).

dump(S, (L,H,R)) :-
    reverse(L, T, [[H]|R]),
    write(T),
    put_char(/),
    put_code(S),
    nl.

reverse([X|Ls], R, A) :-
    reverse(Ls, R, [X|A]).
reverse([], R, R).
```

**Comment by jburse** [comment]
Hi,

Ok, I can reproduce the error:

java.lang.NullPointerException
    at jekpro.model.builtin.SpecialBody.moniFirst(SpecialBody.java:99)
    at jekpro.model.rope.Goal.resolveNext(Goal.java:122)
    at jekpro.model.inter.Engine.runLoop2(Engine.java:195)
    at jekpro.reference.runtime.SpecialSession.sessionTerminal(SpecialSession.java:334)
    at jekpro.reference.runtime.SpecialSession.moniFirst(SpecialSession.java:93)
    at jekpro.model.rope.Goal.resolveNext(Goal.java:122)
    at jekpro.model.inter.Engine.runLoop2(Engine.java:195)
    at jekpro.tools.call.CallIn.unfoldFirst(CallIn.java:363)
    at jekpro.tools.call.CallIn.next(CallIn.java:168)
    at jekpro.util.abortable.ScrollInterpreter.run(ScrollInterpreter.java:292)
    at java.lang.Thread.run(Thread.java:748)

Which has to do with the new disjunction/if-then-else.

**Comment by jburse** [comment]
Hi,

Here is workaround for your code. If you write prepare/3 as follows:

```
parse(_, _, []) :- !.
parse(H, S, [X,RL,S1|Ds]) :-
    H1 is X-0'0, 
    ( RL = 0'R -> B = wsr(H,H1,T,T1) ; 
      RL = 0'L -> B = wsl(H,H1,T,T1) ; 
      fail ),   %%% new
    assertz((trans(S,T,S1,T1) :- B, !)),
    ( H = 0 -> parse(1, S, Ds)
    ; H = 1 -> N is S+1, parse(0, N, Ds)
    ; fail ).   %%% new
```

You don't get some error anymore. Here is a run:

![floaacebpdknbmmm](https://user-images.githubusercontent.com/5411487/57576520-a75c3b00-7461-11e9-823a-6333de949ccf.png)

I will see to it, that the inlining of disjunction/if-then-else
gets more tolerant, so that a missing "; fail" is also tolerated.

Release 1.3.7 should come out in a few days.
Thank you very much for reporting!

Best Regards

**Comment by jburse** [comment]
Fixed with 30497b5a2621b67ad01f9e3938dcf408e8d89a7d
